cpp/src/gandiva/decimal_ir.cc

```diff
@@ -0,0 +1,470 @@
+      auto x_neg = ir_builder()->CreateNeg(x_value);
+      auto y_neg = ir_builder()->CreateNeg(y_value);
+      auto res = AddLargePositive(x_neg, x_scale, y_neg, y_scale, out_scale);
+      return ir_builder()->CreateNeg(res);
+    };
+    return BuildIfElse(
+        x_le_zero_and_y_le_zero, types()->i128_type(), then_lambda_inner,
+        [&] { return AddLargeNegative(x_value, x_scale, y_value, y_scale, out_scale); });
+  };
+  return BuildIfElse(x_ge_zero_and_y_ge_zero, types()->i128_type(), then_lambda,
+                     else_lambda);
+}
+
+/// The output scale/precision cannot be arbitary values. The algo here depends on them
+/// to be the same as computed in DecimalTypeUtil.
+/// TODO: enforce this.
+Status DecimalIR::BuildAdd() {
+  // Create fn prototype :
+  //   void add_decimal128_decimal128(decimal128 *in1, decimal128 *in2, decimal128 *out)
+  std::vector<llvm::Type*> arguments;
+  arguments.push_back(types()->decimal128_ptr_type());
+  arguments.push_back(types()->decimal128_ptr_type());
+  arguments.push_back(types()->decimal128_ptr_type());
+  llvm::FunctionType* prototype =
+      llvm::FunctionType::get(types()->void_type(), arguments, false /*isVarArg*/);
+
+  // Create fn
+  std::string function_name = "add_decimal128_decimal128";
+  function_ = llvm::Function::Create(prototype, llvm::GlobalValue::ExternalLinkage,
+                                     function_name, module());
+  ARROW_RETURN_FAILURE_IF_FALSE((function_ != nullptr),
+                                Status::CodeGenError("Error creating function."));
+  // Name the arguments
+  llvm::Function::arg_iterator args = function_->arg_begin();
+  llvm::Value* arg_x = &*args;
+  arg_x->setName("x");
+  ++args;
+  llvm::Value* arg_y = &*args;
+  arg_y->setName("y");
+  ++args;
+  llvm::Value* arg_out = &*args;
+  arg_out->setName("out");
+  ++args;
+
+  llvm::BasicBlock* entry = llvm::BasicBlock::Create(*context(), "entry", function_);
+  llvm::BasicBlock* exit = llvm::BasicBlock::Create(*context(), "exit", function_);
+
+  ir_builder()->SetInsertPoint(entry);
+
+  // extract members
+  auto x_value = ExtractMemberFromPtr(arg_x, LLVMTypes::kDecimalValueIdx);
+  auto x_scale = ExtractMemberFromPtr(arg_x, LLVMTypes::kDecimalScaleIdx);
+  auto y_value = ExtractMemberFromPtr(arg_y, LLVMTypes::kDecimalValueIdx);
+  auto y_scale = ExtractMemberFromPtr(arg_y, LLVMTypes::kDecimalScaleIdx);
+  auto out_scale = ExtractMemberFromPtr(arg_out, LLVMTypes::kDecimalScaleIdx);
+  auto out_precision = ExtractMemberFromPtr(arg_out, LLVMTypes::kDecimalPrecisionIdx);
+
+  llvm::Value* lt_max_precision = ir_builder()->CreateICmpSLT(
+      out_precision, types()->i32_constant(DecimalTypeUtil::kMaxPrecision));
+  auto then_lambda = [&] {
+    // fast-path add
+    return AddFastPath(x_value, x_scale, y_value, y_scale);
+  };
+  auto else_lambda = [&] {
+    // slower version : add whole/fraction parts separately
```

Comment:
I don't understand this: when is the slow path needed? When you have full precision?

Reply:
yes - when, the result precision is = 38.